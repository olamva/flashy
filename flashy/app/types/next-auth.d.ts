import "next-auth";
import { User as UserType } from "./user";

declare module "next-auth" {
  interface Session {
    user: UserType & Session["user"];
    expires: Session["expires"]
  }

  interface User extends DefaultUser {
    id: UserType["id"];
    name: UserType["name"];
    email: UserType["email"];
    image: UserType["image"];
    role: UserType["role"];
    username: UserType["username"];
    age: UserType["age"];
  }
}