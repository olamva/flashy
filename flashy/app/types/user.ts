export type User = {
    id: string;
    email: string;
    name: string;
    image: string;
    role: string;
    age: number | null;
    username: string;
}