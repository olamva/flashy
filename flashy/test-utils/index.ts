export * from "@testing-library/react";
export * from "./mock";
export { render } from "./render";
