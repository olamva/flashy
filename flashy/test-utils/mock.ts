// Overrides/Mocks router to prevent errors in tests
jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
    };
  },
  usePathname() {
    return "/";
  },
}));

// Mocks useSession hook
// jest.mock("next-auth/react", () => ({
//     useSession: jest.fn(() => ({ data: { user: dummyAdmin } }))
// }));

export {};
