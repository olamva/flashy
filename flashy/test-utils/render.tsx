import { MantineProvider } from '@mantine/core';
import { render as testingLibraryRender } from '@testing-library/react';

// From official Mantine documentation
export function render(ui: React.ReactNode) {
    return testingLibraryRender(<>{ui}</>, {
        wrapper: ({ children }: { children: React.ReactNode }) => (
            <MantineProvider>{children}</MantineProvider>
        ),
    });
}