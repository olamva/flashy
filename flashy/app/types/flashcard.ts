import { User } from "./user";

export enum Visibility {
  Public = "Offentlig",
  Private = "Privat",
}

export type FlashcardSet = {
  id: string;
  creator?: User;
  coAuthors?: User[];
  coverImage?: string | File;
  title: string;
  numViews: number;
  numOfLikes: number;
  numOfFavorites: number;
  numOfComments?: number;
  userHasLiked?: boolean;
  userHasFavorited?: boolean;
  comments?: FlashcardComment[];
  views?: FlashcardView[];
  visibility?: Visibility;
  createdAt: Date;
  popularityScore: number;
};

// Used to save on read operations
export type ShallowFlashcardSet = {
  id: string;
  creator: User | undefined;
  title: string;
  numViews: number;
  numOfLikes: number;
  visibility: Visibility;
};

export type FlashcardFlagged = {
  cardsFlagged: string[];
};

export type FlashcardComment = {
  id: string;
  commentedBy: User | undefined;
  content: string;
  createdAt: Date;
};

export type FlashcardView = {
  id: string;
  isCopy?: boolean;
  front: string;
  back: string;
  image?: string;
};

export type CreateFlashCardViewType = FlashcardView & {
  image?: File;
};

export type CreateFlashCardType = {
  creator: User;
  coAuthors: string[];
  title: string;
  views: CreateViewType[];
  visibility: Visibility;
  createdAt: Date;
  image?: File;
};

export type CreateViewType = Pick<CreateFlashCardViewType, "front" | "back" | "image">;

export type EditFlashCardType = {
  views: EditFlashcardView[];
  visibility: Visibility;
  coAuthors: string[];
  coverImage: File | undefined;
};

export type EditFlashcardView = Pick<FlashcardView, "front" | "back" | "image"> & {
  id?: string;
};

export type CreateNewCommentType = {
  commentedBy: User;
  content: string;
};
