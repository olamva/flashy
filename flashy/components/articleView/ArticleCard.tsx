import { FlashcardSet } from "@/app/types/flashcard";
import { setIncrementFlashcardViews } from "@/app/utils/firebase";
import classes from "@/components/articleView/ArticleCardsGrid.module.css";
import { AspectRatio, Card, Container, Divider, Group, Space, Stack, Text, UnstyledButton } from "@mantine/core";
import { IconBrandHipchat, IconEyeUp, IconFriends, IconThumbUp } from "@tabler/icons-react";
import { Session } from "next-auth";
import Image from "next/image";
import { useRouter } from "next/navigation";

type ArticleCardProps = {
  flashcard: FlashcardSet;
  user: Session["user"];
};

export function ArticleCard({ flashcard }: ArticleCardProps) {
  const router = useRouter();

  const handleClickCarousel = () => {
    setIncrementFlashcardViews(flashcard);
    router.push("/carousel/" + flashcard.title);
  };

  return (
    <Card key={flashcard.title} p="0" shadow="sm" radius="md" component="a" className={classes.card}>
      <UnstyledButton p="md" onClick={() => handleClickCarousel()}>
        <AspectRatio ratio={1920 / 1080}>
          <Image
            style={{ borderRadius: 5 }}
            src={typeof flashcard.coverImage === "string" && flashcard.coverImage != "" ? flashcard.coverImage : "https://picsum.photos/300?grayscale"}
            alt="Forsidebilde for Flashy"
            width={200}
            height={100}
            priority={true}
          />
        </AspectRatio>
        <Space h={10} />
        <Stack align="left" w={"100%"}>
          <Stack gap={0}>
            <Text c="dimmed" size="xs" tt="uppercase" fw={700}>
              üë®‚Äçüé® {flashcard.creator?.name ?? "slettet bruker"}{" "}
              {flashcard.coAuthors?.length ? (
                <>
                  <Text style={{ display: "inline" }}> +{flashcard.coAuthors.length}</Text>
                  <IconFriends width={15} style={{ verticalAlign: "middle" }}></IconFriends>
                </>
              ) : (
                <></>
              )}
            </Text>
            <Text c="dimmed" tt="capitalize" fw={600} size="xl">
              {flashcard.title}
            </Text>
          </Stack>
          <Group justify="space-between" w={"100%"} style={{ display: "flex" }}>
            <Container style={{ display: "flex", alignItems: "center" }}>
              <Text size="xl" fw={700} style={{ paddingRight: 2 }}>
                {flashcard.numOfLikes ?? 0}
              </Text>
              <IconThumbUp />
            </Container>
            <Divider orientation="vertical" />
            <Container style={{ display: "flex", alignItems: "center" }}>
              <Text size="xl" fw={700} style={{ paddingRight: 2 }}>
                {flashcard.numViews ?? 0}
              </Text>
              <IconEyeUp />
            </Container>
            <Divider orientation="vertical" />
            <Container style={{ display: "flex", alignItems: "center" }}>
              <Text size="xl" fw={700} style={{ paddingRight: 2 }}>
                {flashcard.numOfComments ?? 0}
              </Text>
              <IconBrandHipchat />
            </Container>
          </Group>
        </Stack>
      </UnstyledButton>
    </Card>
  );
}
