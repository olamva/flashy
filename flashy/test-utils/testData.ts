import { FlashcardSet, FlashcardView } from "@/app/types/flashcard"
import { User } from "next-auth"

export const dummyFlashcard: FlashcardSet = {
  id: "1",
  title: "dummy",
  numViews: 2,
  numOfLikes: 2,
  numOfFavorites: 4,
  numOfComments: 3,
  createdAt: new Date(),
  popularityScore: 2,
}

export const dummyUser: User = {
  id: 'abc',
  email: "test@test.com",
  name: "Ola Nordmann",
  image: "",
  role: "user",
  age: 20,
  username: "olanordmann",
}

export const dummyAdmin: User = {
  id: 'abcd',
  email: "test@test.com",
  name: "Ola Nordmann Admin",
  image: "",
  role: "admin",
  age: 20,
  username: "olanordmannadmin",
}

export const dummyViews: FlashcardView[] = [
  {
    id: "1",
    front: "Front1",
    back: "Back1",
  },
  {
    id: "2",
    front: "Front2",
    back: "Back2",
  },
  {
    id: "3",
    front: "Front3",
    back: "Back3",
  },
]