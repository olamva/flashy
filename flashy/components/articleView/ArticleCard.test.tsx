import { FlashcardSet } from '@/app/types/flashcard';
import { render, screen } from '@/test-utils';
import { dummyFlashcard, dummyUser } from '@/test-utils/testData';
import { ArticleCard } from './ArticleCard';

function countOccurences(flashcardSet: FlashcardSet): { [key: number]: number } {
    const [numOfLikes, numViews, numOfComments] = [flashcardSet.numOfLikes, flashcardSet.numViews, flashcardSet.numOfComments ?? 0];

    const occurrences: { [key: number]: number } = {};

    [numOfLikes, numViews, numOfComments].forEach((value) => {
        if (occurrences[value]) {
            occurrences[value]++;
        } else {
            occurrences[value] = 1;
        }
    });

    return occurrences;
}

describe("ArticleCard Component Test", () => {
    it("renders correctly", () => {
        render(<ArticleCard flashcard={dummyFlashcard} user={dummyUser} />);

        // Checks whether title is displayed
        expect(screen.getByText(dummyFlashcard.title)).toBeInTheDocument();

        // Checks whether information displayed is correct
        const occurrences = countOccurences(dummyFlashcard);
        expect(screen.getAllByText(dummyFlashcard.numViews)).toHaveLength(occurrences[dummyFlashcard.numViews]);
        expect(screen.getAllByText(dummyFlashcard.numOfLikes)).toHaveLength(occurrences[dummyFlashcard.numOfLikes]);
        expect(screen.getAllByText(dummyFlashcard.numOfComments ?? 0)).toHaveLength(occurrences[dummyFlashcard.numOfComments ?? 0]);
    });
});

describe("ArticleCard Snapshot Test", () => {
    it("Snapshot matches", () => {
        const { asFragment } = render(<ArticleCard flashcard={dummyFlashcard} user={dummyUser} />);;
        expect(asFragment()).toMatchSnapshot();
    });
});