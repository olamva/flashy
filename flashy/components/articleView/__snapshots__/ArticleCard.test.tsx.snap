// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard Snapshot Test Snapshot matches 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <a
    class="card m-e615b15f mantine-Card-root m-1b7284a3 mantine-Paper-root"
    style="--paper-radius: var(--mantine-radius-md); --paper-shadow: var(--mantine-shadow-sm); padding: 0rem;"
  >
    <button
      class="mantine-focus-auto m-87cf2631 mantine-UnstyledButton-root"
      type="button"
    >
      <div
        class="m-71ac47fc mantine-AspectRatio-root"
        style="--ar-ratio: 1.7777777777777777;"
      >
        <img
          alt="Forsidebilde for Flashy"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="100"
          src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F300%3Fgrayscale&w=640&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F300%3Fgrayscale&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F300%3Fgrayscale&w=640&q=75 2x"
          style="color: transparent; border-radius: 5px;"
          width="200"
        />
      </div>
      <div
        class=""
        style="min-height: calc(0.625rem * var(--mantine-scale));"
      />
      <div
        class="m-6d731127 mantine-Stack-root"
        style="--stack-gap: var(--mantine-spacing-md); --stack-align: left; --stack-justify: flex-start; width: 100%;"
      >
        <div
          class="m-6d731127 mantine-Stack-root"
          style="--stack-gap: 0rem; --stack-align: stretch; --stack-justify: flex-start;"
        >
          <p
            class="mantine-focus-auto m-b6d8b162 mantine-Text-root"
            data-size="xs"
            style="--text-fz: var(--mantine-font-size-xs); --text-lh: var(--mantine-line-height-xs); font-weight: 700; text-transform: uppercase;"
          >
            üë®‚Äçüé® slettet bruker 
          </p>
          <p
            class="mantine-focus-auto m-b6d8b162 mantine-Text-root"
            data-size="xl"
            style="--text-fz: var(--mantine-font-size-xl); --text-lh: var(--mantine-line-height-xl); font-weight: 600; text-transform: capitalize;"
          >
            dummy
          </p>
        </div>
        <div
          class="m-4081bf90 mantine-Group-root"
          style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: space-between; --group-wrap: wrap; display: flex; width: 100%;"
        >
          <div
            class="m-7485cace mantine-Container-root"
            style="display: flex; align-items: center;"
          >
            <p
              class="mantine-focus-auto m-b6d8b162 mantine-Text-root"
              data-size="xl"
              style="--text-fz: var(--mantine-font-size-xl); --text-lh: var(--mantine-line-height-xl); padding-right: 2px; font-weight: 700;"
            >
              2
            </p>
            <svg
              class="tabler-icon tabler-icon-thumb-up"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 11v8a1 1 0 0 1 -1 1h-2a1 1 0 0 1 -1 -1v-7a1 1 0 0 1 1 -1h3a4 4 0 0 0 4 -4v-1a2 2 0 0 1 4 0v5h3a2 2 0 0 1 2 2l-1 5a2 3 0 0 1 -2 2h-7a3 3 0 0 1 -3 -3"
              />
            </svg>
          </div>
          <div
            class="m-3eebeb36 mantine-Divider-root"
            data-orientation="vertical"
            role="separator"
          />
          <div
            class="m-7485cace mantine-Container-root"
            style="display: flex; align-items: center;"
          >
            <p
              class="mantine-focus-auto m-b6d8b162 mantine-Text-root"
              data-size="xl"
              style="--text-fz: var(--mantine-font-size-xl); --text-lh: var(--mantine-line-height-xl); padding-right: 2px; font-weight: 700;"
            >
              2
            </p>
            <svg
              class="tabler-icon tabler-icon-eye-up"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0"
              />
              <path
                d="M12 18c-3.6 0 -6.6 -2 -9 -6c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.09 .15 -.18 .295 -.27 .439"
              />
              <path
                d="M19 22v-6"
              />
              <path
                d="M22 19l-3 -3l-3 3"
              />
            </svg>
          </div>
          <div
            class="m-3eebeb36 mantine-Divider-root"
            data-orientation="vertical"
            role="separator"
          />
          <div
            class="m-7485cace mantine-Container-root"
            style="display: flex; align-items: center;"
          >
            <p
              class="mantine-focus-auto m-b6d8b162 mantine-Text-root"
              data-size="xl"
              style="--text-fz: var(--mantine-font-size-xl); --text-lh: var(--mantine-line-height-xl); padding-right: 2px; font-weight: 700;"
            >
              3
            </p>
            <svg
              class="tabler-icon tabler-icon-brand-hipchat"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17.802 17.292s.077 -.055 .2 -.149c1.843 -1.425 3 -3.49 3 -5.789c0 -4.286 -4.03 -7.764 -9 -7.764c-4.97 0 -9 3.478 -9 7.764c0 4.288 4.03 7.646 9 7.646c.424 0 1.12 -.028 2.088 -.084c1.262 .82 3.104 1.493 4.716 1.493c.499 0 .734 -.41 .414 -.828c-.486 -.596 -1.156 -1.551 -1.416 -2.29z"
              />
              <path
                d="M7.5 13.5c2.5 2.5 6.5 2.5 9 0"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </a>
</DocumentFragment>
`;
